map_model/src/pathfind/v2.rs

```diff
@@ -64,6 +70,49 @@ pub fn path_v2_to_v1(req: PathRequest, road_steps: Vec<DirectedRoadID>, map: &Ma
     }
 }
 
+fn find_uber_turns(
+    steps: &Vec<PathStep>,
+    map: &Map,
+    mut uber_turns_v2: Vec<UberTurnV2>,
+) -> Vec<UberTurn> {
+    // Pathfinding v1 needs to know the uber turns that the path crosses, for the simulation layer.
+    // Since we now construct the path in two stages, it's easiest to just reconstruct the uber
+    // turns after building the lane-based path.
```

Comment:
> Since we now construct the path in two stages

Here's my understanding:

> we now construct the path in two stages

Once the pathfinding v2 stuff is complete, uber-turns will still be relevant, but not during path finding, which is the first stage of path construction, where Lanes are irrelevant, we're only concerned about Roads.

The second stage of path construction does know about Lanes. It is responsible for initial lane assignment and intersection traversal/dynamic lane changing. We will need to know about uber turns for this stage.

Is that right?

Reply:
This is correct. I won't know for sure until I go through all callers of the current pathfinding API, but I believe most of them just need to estimate path length, draw the path, or figure out if the path even exists at all. The only one I know of that needs the full detail of lanes is the simulation. So my plan is to have most places use a new pathfinding v2 API (that I've yet to write) that stays at the level of `DirectedRoadID`s. The first cut of the simulation would be to call this `path_v2_to_v1` function.

But we can do much better -- like storing the v2 path (road segments) and lazily calculating the next few lanes and turns as needed inside of `Router` and the equivalent for pedestrians. That lazy evaluation place can also discover the uber-turns as needed. I don't think preserving the uber turns from the road-based graph is necessary or useful; we can cache the lanes acting as uber-turn entrances and exits somewhere, and the lazy path evaluation can just do a lookup there. We use uber-turns at the sim level to reserve a bunch of turns in advance, run red lights, and fiddle with blocking-the-box behavior. It's also possible some of that logic won't last forever, and that we'll robustify intersection merging or something else.

To recap, uber-turns have two different purposes. At the map layer, they encode OSM turn restriction relations that're multiple road segments long -- things like no U-turn, where the middle segments are mapped as distinct OSM ways. The graph used for pathfinding has to be aware of this to avoid returning illegal paths. The second purpose is at the simulation layer -- we use uber-turns as a hint for making it past a cluster of intersections successfully.

Reply:
> Once the pathfinding v2 stuff is complete, uber-turns will still be relevant, but not during path finding

Ah sorry, this part is incorrect. The first purpose, "encode OSM turn restriction relations", is still true with pathfinding v2.

Reply:
https://www.openstreetmap.org/relation/12384357#map=19/47.62350/-122.29639 is an example of one of those OSM turn restriction relations

Reply:
Got it! That all makes sense. Thanks for the explanation.